File: examples/migration.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use iota_wallet::{
    account_manager::{AccountManager, MigrationDataFinder},
    client::ClientOptionsBuilder,
    signing::SignerType,
};
use std::time::Duration;

#[tokio::main]
async fn main() -> iota_wallet::Result<()> {
    let manager = AccountManager::builder().finish().await.unwrap();
    manager.set_stronghold_password("password").await.unwrap();
    manager.store_mnemonic(SignerType::Stronghold, None).await.unwrap();

    // first we'll create an example account and store it
    let client_options = ClientOptionsBuilder::new()
        .with_node("https://api.lb-0.h.chrysalis-devnet.iota.cafe")?
        .build()
        .unwrap();
    let _account = manager
        .create_account(client_options)?
        .alias("alias")
        .initialise()
        .await?;

    // Migration
    let legacy_node = "https://nodes.devnet.iota.org";
    let seed = "TRYTESEEDTRYTESEEDTRYTESEEDTRYTESEEDTRYTESEEDTRYTESEEDTRYTESEEDTRYTESEEDTRYTESEED";
    let min_weight_magnitude = 9;

    // Get account data
    let mut address_index = 0;
    let gap_limit = 30;
    let yes = vec!['Y', 'y'];
    let mut user_input = String::new();
    let mut migration_data = None;
    while !yes.contains(&user_input.chars().next().unwrap_or('N')) {
        let account_migration_data = manager
            .get_migration_data(
                MigrationDataFinder::new(&[legacy_node], seed)?
                    .with_initial_address_index(address_index)
                    .with_gap_limit(gap_limit)
                    .with_permanode("https://chronicle.iota.org/api"),
            )
            .await?;
        println!(
            "Is {}i the correct balance? Type Y to continue or N to search for more balance",
            account_migration_data.balance
        );
        user_input = String::new();
        std::io::stdin().read_line(&mut user_input).unwrap();
        address_index += gap_limit;
        migration_data = Some(account_migration_data);
    }

    let migration_data = migration_data.unwrap();

    let mut bundle_hashes = Vec::new();
    for input in migration_data.inputs.into_iter() {
        let bundle = manager
            .create_migration_bundle(
                seed,
                &[input.index],
                true,
                Duration::from_secs(40),
                0,
                "./migration.log",
            )
            .await?;
        bundle_hashes.push(bundle.bundle_hash().to_string());
    }

    for hash in bundle_hashes {
        manager
            .send_migration_bundle(&[legacy_node], &hash, min_weight_magnitude)
            .await?;
        println!("Bundle sent, hash: {}", hash);
    }

    Ok(())
}

File: bindings/python/examples/migration.py
# Copyright 2020 IOTA Stiftung
# SPDX-License-Identifier: Apache-2.0


import iota_wallet as iw
import os


# Get the stronghold password, please get the password from your environment
# NEVER assign your password directly in the code!
STRONGHOLD_PASSWORD = "UNSAFEPASSWORD"

# note: `storage` and `storage_path` have to be declared together
account_manager = iw.AccountManager(
    storage_path='./alice-database'
)

account_manager.set_stronghold_password(STRONGHOLD_PASSWORD)

# mnemonic (seed) should be set only for new storage
# once the storage has been initialized earlier then you should omit this step
account_manager.store_mnemonic("Stronghold")

# General Tangle specific options
client_options = {
    "primary_node": None,
    "primary_pow_node": None,
    "nodes": [
        {
            "url": "https://api.lb-0.h.chrysalis-devnet.iota.cafe",
            "auth": None,
            "disabled": False
        }
    ],
    "local_pow": None,
    "request_timeout": 60000,
    "api_timeout": {
        "PostMessage": 3000
    }
}

# An account is generated with the given alias via `account_initializer`
account_initializer = account_manager.create_account(client_options)
account_initializer.alias('Alice')

# Initialise account based via `initializer`
# Store it to db and sync with Tangle
account = account_initializer.initialise()
print(f'Account created: {account.alias()}')

# Migration
legacy_node = 'https://nodes.devnet.iota.org'
seed = 'TRYTESEEDTRYTESEEDTRYTESEEDTRYTESEEDTRYTESEEDTRYTESEEDTRYTESEEDTRYTESEEDTRYTESEED'
min_weight_magnitude = 9
permanode = 'https://chronicle.iota.org/api'


# Get the account data
address_index = 0
while True:
    migration_data = account_manager.get_migration_data(
        initial_address_index=address_index, nodes=[legacy_node], seed=seed, permanode=permanode)

    address_index += 30
    print(f'Is {migration_data["balance"]}i the correct balance?')
    print('Type N to search for more balance or type any other keys to exit.')
    if input() != 'N':
        break

# Send the migration bundle
input_address_indexes = []
for input in migration_data['inputs']:
    input_address_indexes.append(input['index'])
    try:
        bundle = account_manager.create_migration_bundle(
            seed, input_address_indexes, mine=True, timeout_seconds=40, offset=0)

        account_manager.send_migration_bundle(
            [legacy_node], bundle['bundle_hash'], min_weight_magnitude)

        hash = bundle['bundle_hash']
        print(f'Bundle sent, hash: {hash}')
    except ValueError as e:
        print(f'Error! {e}')
